massa-bootstrap/src/listener.rs

```diff
@@ -6,69 +6,79 @@ use mio::{Events, Interest, Poll, Token, Waker};
 use tracing::{error, info};
 
 use crate::error::BootstrapError;
+use crate::establisher::BSListener;
 
 const NEW_CONNECTION: Token = Token(0);
 const STOP_LISTENER: Token = Token(10);
 
-pub(crate) struct BootstrapTcpListener();
+/// TODO: this should be crate-private. currently needed for models testing
+pub struct BootstrapTcpListener {
+    poll: Poll,
+    events: Events,
+    server: TcpListener,
+    // HACK : create variable to move ownership of mio_server to the thread
+    // if mio_server is not moved, poll does not receive any event from listener
+    _mio_server: MioTcpListener,
+}
 
 pub struct BootstrapListenerStopHandle(Waker);
 
+pub enum AcceptEvent {
+    NewConnection((TcpStream, SocketAddr)),
+    Stop,
+}
 impl BootstrapTcpListener {
     /// Start a new bootstrap listener on the given address.
     ///
     /// * `addr` - the address to listen on
     /// * `connection_tx` - the channel to send new connections to
-    pub fn start(
-        addr: SocketAddr,
-        connection_tx: crossbeam::channel::Sender<(TcpStream, SocketAddr)>,
-    ) -> Result<BootstrapListenerStopHandle, BootstrapError> {
+    pub fn new(addr: SocketAddr) -> Result<(BootstrapListenerStopHandle, Self), BootstrapError> {
         info!("Starting bootstrap listener on {}", &addr);
         let server = TcpListener::bind(addr)?;
         let mut mio_server =
             MioTcpListener::from_std(server.try_clone().expect("Unable to clone server socket"));
 
-        let mut poll = Poll::new()?;
+        let poll = Poll::new()?;
 
         // wake up the poll when we want to stop the listener
-        let waker = Waker::new(poll.registry(), STOP_LISTENER)?;
+        let waker = BootstrapListenerStopHandle(Waker::new(poll.registry(), STOP_LISTENER)?);
 
         poll.registry()
             .register(&mut mio_server, NEW_CONNECTION, Interest::READABLE)?;
 
         // TODO use config for capacity ?
-        let mut events = Events::with_capacity(32);
-
-        // spawn a thread to handle events
-        std::thread::Builder::new()
-            .name("bs_listener".to_string())
-            .spawn(move || {
-                // HACK : create variable to move ownership of mio_server to the thread
-                // if mio_server is not moved, poll does not receive any event from listener
-                let _temp = mio_server;
+        let events = Events::with_capacity(32);
+        Ok((
+            waker,
+            BootstrapTcpListener {
+                poll,
+                server,
+                events,
+                _mio_server: mio_server,
+            },
+        ))
+    }
+}
```

Comment:
The intent here is to provide the server itself the ability to call a blocking-accept in the main-loop, and be able to determine if it needs to stop